dids/did_dht_test.go

```diff
@@ -0,0 +1,170 @@
+		didURI        string
+		msg           dnsmessage.Message
+		expectedError error
+		assertResult  func(t *testing.T, d *Document)
+	}{
+		"did with valid key and no service": {
+			didURI: "did:dht:cwxob5rbhhu3z9x3gfqy6cthqgm6ngrh4k8s615n7pw11czoq4fy",
+			msg: makeDNSMessage(
+				WithDNSRecord("_did.", "vm=k0;auth=k0;asm=k0;inv=k0;del=k0"),
+				WithDNSRecord("_k0._did.", "id=0;t=0;k=YCcHYL2sYNPDlKaALcEmll2HHyT968M4UWbr-9CFGWE"),
+			),
+			expectedError: nil,
+			assertResult: func(t *testing.T, d *Document) {
+				assert.False(t, d == nil, "Expected non nil document")
+				assert.NotZero[string](t, d.ID, "Expected DID Document ID to be initialized")
+				assert.NotZero[[]VerificationMethod](t, d.VerificationMethod, "Expected at least 1 verification method")
+			},
+		},
+		"did with multiple valid keys and no service - out of order verification methods": {
+			didURI: "did:dht:cwxob5rbhhu3z9x3gfqy6cthqgm6ngrh4k8s615n7pw11czoq4fy",
+			msg: makeDNSMessage(
+				WithDNSRecord("_did.", "vm=k0,k1,k2;auth=k0;asm=k1;inv=k2;del=k0"),
+				WithDNSRecord("_k0._did.", "id=0;t=0;k=YCcHYL2sYNPDlKaALcEmll2HHyT968M4UWbr-9CFGWE"),
+				WithDNSRecord("_k2._did.", fmt.Sprintf("id=2;t=1;k=%s", base64EncodedSecp256k)),
+				WithDNSRecord("_k1._did.", fmt.Sprintf("id=1;t=2;k=%s", base64EncodedSecp256k)),
+			),
+			expectedError: nil,
+			assertResult: func(t *testing.T, d *Document) {
+				assert.False(t, d == nil, "Expected non nil document")
+				assert.NotZero[string](t, d.ID, "Expected DID Document ID to be initialized")
+				assert.Equal[int](t, 3, len(d.VerificationMethod), "Expected 3 verification methods")
+			},
+		},
+		"did with key controller and services": {
+			didURI: "did:dht:cwxob5rbhhu3z9x3gfqy6cthqgm6ngrh4k8s615n7pw11czoq4fy",
+			msg: makeDNSMessage(
+				WithDNSRecord("_did.", "vm=k0;auth=k0;asm=k0;inv=k0;del=k0;srv=s0,s1"),
+				WithDNSRecord("_k0._did.", "id=0;t=0;k=YCcHYL2sYNPDlKaALcEmll2HHyT968M4UWbr-9CFGWE"),
+				WithDNSRecord("_s0._did.", "id=domain;t=LinkedDomains;se=foo.com"),
+				WithDNSRecord("_s1._did.", "id=dwn;t=DecentralizedWebNode;se=https://dwn.tbddev.org/dwn5"),
+			),
+			expectedError: nil,
+			assertResult: func(t *testing.T, d *Document) {
+				assert.False(t, d == nil, "Expected non nil document")
+				assert.NotZero[string](t, d.ID, "Expected DID Document ID to be initialized")
+				assert.NotZero[[]VerificationMethod](t, d.VerificationMethod, "Expected at least 1 verification method")
+				assert.Equal[int](t, 2, len(d.Service), "Expected 2 services")
+			},
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			buf, err := test.msg.Pack()
+			assert.NoError(t, err)
+			// test setup
+			ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				_, err := w.Write(buf)
+				assert.NoError(t, err)
+			}))
+			defer ts.Close()
+
+			result, err := ResolveDIDDHT(test.didURI, ts.URL, http.DefaultClient)
+
+			if test.expectedError != nil {
```

Comment:
Could have `expectedError` be a string, then use [assert.EqualError()](https://pkg.go.dev/github.com/alecthomas/assert/v2#EqualError) here. It will eliminate this if block.

Reply:
nice